// astro/dist/core/build/page-data.js
import { fileURLToPath } from "url";
import * as colors from "kleur/colors";
import { debug } from "../logger.js";
import { preload as ssrPreload } from "../render/dev/index.js";
import { generateRssFunction } from "../render/rss.js";
import { callGetStaticPaths } from "../render/route-cache.js";
async function collectPagesData(opts) {
  const { astroConfig, logging, manifest, origin, routeCache, viteServer } = opts;
  const assets = {};
  const allPages = {};
  await Promise.all(manifest.routes.map(async (route) => {
    var _a;
    if (route.pathname) {
      allPages[route.component] = {
        route,
        paths: [route.pathname],
        preload: await ssrPreload({
          astroConfig,
          filePath: new URL(`./${route.component}`, astroConfig.projectRoot),
          logging,
          mode: "production",
          origin,
          pathname: route.pathname,
          route,
          routeCache,
          viteServer
        }).then((routes) => {
          const html = `${route.pathname}`.replace(/\/?$/, "/index.html");
          debug("build", `\u251C\u2500\u2500 ${colors.bold(colors.green("\u2714"))} ${route.component} \u2192 ${colors.yellow(html)}`);
          return routes;
        }).catch((err) => {
          debug("build", `\u251C\u2500\u2500 ${colors.bold(colors.red("\u2718"))} ${route.component}`);
          throw err;
        })
      };
      return;
    }
    const result = await getStaticPathsForRoute(opts, route).then((_result) => {
      const label = _result.staticPaths.length === 1 ? "page" : "pages";
      debug("build", `\u251C\u2500\u2500 ${colors.bold(colors.green("\u2714"))} ${route.component} \u2192 ${colors.magenta(`[${_result.staticPaths.length} ${label}]`)}`);
      return _result;
    }).catch((err) => {
      debug("build", `\u251C\u2500\u2500 ${colors.bold(colors.red("\u2717"))} ${route.component}`);
      throw err;
    });
    const rssFn = generateRssFunction(astroConfig.buildOptions.site, route);
    for (const rssCallArg of result.rss) {
      const rssResult = rssFn(rssCallArg);
      if (rssResult.xml) {
        const { url, content } = rssResult.xml;
        if (content) {
          const rssFile = new URL(url.replace(/^\/?/, "./"), astroConfig.dist);
          if (assets[fileURLToPath(rssFile)]) {
            throw new Error(`[getStaticPaths] RSS feed ${url} already exists.
Use \`rss(data, {url: '...'})\` to choose a unique, custom URL. (${route.component})`);
          }
          assets[fileURLToPath(rssFile)] = content;
        }
      }
      if ((_a = rssResult.xsl) == null ? void 0 : _a.content) {
        const { url, content } = rssResult.xsl;
        const stylesheetFile = new URL(url.replace(/^\/?/, "./"), astroConfig.dist);
        if (assets[fileURLToPath(stylesheetFile)]) {
          throw new Error(`[getStaticPaths] RSS feed stylesheet ${url} already exists.
Use \`rss(data, {stylesheet: '...'})\` to choose a unique, custom URL. (${route.component})`);
        }
        assets[fileURLToPath(stylesheetFile)] = content;
      }
    }
    const finalPaths = result.staticPaths.map((staticPath) => staticPath.params && route.generate(staticPath.params)).filter(Boolean);
    allPages[route.component] = {
      route,
      paths: finalPaths,
      preload: await ssrPreload({
        astroConfig,
        filePath: new URL(`./${route.component}`, astroConfig.projectRoot),
        logging,
        mode: "production",
        origin,
        pathname: finalPaths[0],
        route,
        routeCache,
        viteServer
      })
    };
  }));
  return { assets, allPages };
}
async function getStaticPathsForRoute(opts, route) {
  const { astroConfig, logging, routeCache, viteServer } = opts;
  if (!viteServer)
    throw new Error(`vite.createServer() not called!`);
  const filePath = new URL(`./${route.component}`, astroConfig.projectRoot);
  const mod = await viteServer.ssrLoadModule(fileURLToPath(filePath));
  const result = await callGetStaticPaths(mod, route, false, logging);
  routeCache.set(route, result);
  return result;
}
export {
  collectPagesData
};
