// astro/dist/core/create-vite.js
import { builtinModules } from "module";
import { fileURLToPath } from "url";
import vite from "./vite.js";
import astroVitePlugin from "../vite-plugin-astro/index.js";
import astroViteServerPlugin from "../vite-plugin-astro-server/index.js";
import astroPostprocessVitePlugin from "../vite-plugin-astro-postprocess/index.js";
import configAliasVitePlugin from "../vite-plugin-config-alias/index.js";
import markdownVitePlugin from "../vite-plugin-markdown/index.js";
import jsxVitePlugin from "../vite-plugin-jsx/index.js";
import envVitePlugin from "../vite-plugin-env/index.js";
import { resolveDependency } from "./util.js";
const ALWAYS_EXTERNAL = new Set([
  ...builtinModules.map((name) => `node:${name}`),
  "@sveltejs/vite-plugin-svelte",
  "micromark-util-events-to-acorn",
  "serialize-javascript",
  "node-fetch",
  "prismjs",
  "shiki",
  "shorthash",
  "unified",
  "whatwg-url"
]);
const ALWAYS_NOEXTERNAL = new Set([
  "astro"
]);
async function createVite(inlineConfig, { astroConfig, logging, mode }) {
  let viteConfig = {
    cacheDir: fileURLToPath(new URL("./node_modules/.vite/", astroConfig.projectRoot)),
    clearScreen: false,
    logLevel: "error",
    optimizeDeps: {
      entries: ["src/**/*"]
    },
    plugins: [
      configAliasVitePlugin({ config: astroConfig }),
      astroVitePlugin({ config: astroConfig, logging }),
      mode === "dev" && astroViteServerPlugin({ config: astroConfig, logging }),
      envVitePlugin({ config: astroConfig }),
      markdownVitePlugin({ config: astroConfig }),
      jsxVitePlugin({ config: astroConfig, logging }),
      astroPostprocessVitePlugin({ config: astroConfig })
    ],
    publicDir: fileURLToPath(astroConfig.public),
    root: fileURLToPath(astroConfig.projectRoot),
    envPrefix: "PUBLIC_",
    server: {
      force: true,
      hmr: process.env.NODE_ENV === "test" || process.env.NODE_ENV === "production" ? false : void 0,
      proxy: {}
    },
    ssr: {
      external: [...ALWAYS_EXTERNAL],
      noExternal: [...ALWAYS_NOEXTERNAL]
    }
  };
  for (const name of astroConfig.renderers) {
    try {
      const { default: renderer } = await import(resolveDependency(name, astroConfig));
      if (!renderer)
        continue;
      if (renderer.viteConfig) {
        if (typeof renderer.viteConfig !== "function") {
          throw new Error(`${name}: viteConfig(options) must be a function! Got ${typeof renderer.viteConfig}.`);
        }
        const rendererConfig = await renderer.viteConfig({ mode: inlineConfig.mode, command: inlineConfig.mode === "production" ? "build" : "serve" });
        viteConfig = vite.mergeConfig(viteConfig, rendererConfig);
      }
    } catch (err) {
      throw new Error(`${name}: ${err}`);
    }
  }
  viteConfig = vite.mergeConfig(viteConfig, inlineConfig);
  return viteConfig;
}
export {
  createVite
};
